[tp] add clarification to doc and improve TP examples

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #117618

This PR adds a clarification about evenly sharded assumption in the main
tp doc and improved the tp examples by adding device mesh constructions

fixes https://github.com/pytorch/pytorch/issues/100044

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @fduwjj @wz337 @tianyu-l @wconstab @yf225

wanchaol committed:
[tp] add clarification to doc and improve TP examples

This PR adds a clarification about evenly sharded assumption in the main
tp doc and improved the tp examples by adding device mesh constructions

fixes https://github.com/pytorch/pytorch/issues/100044

[ghstack-poisoned]

wanchaol commented on docs/source/distributed.tensor.parallel.rst:
```diff
@@ -42,6 +42,11 @@ the ``parallelize_plan`` of ``parallelize_module``:
+  ParallelStyles, we assume the input/output activation tensors are evenly sharded on
+  the tensor dimension ``dim`` on the DeviceMesh that TP operates on. For instance,
+  ``RowwiseParallel`` accepts input that is sharded on the last dimension, it assumes
+  the input tensor already been evenly sharded on the last dimension.
```
@awgu I added some clarification about uneven sharding in the doc

awgu replied:
nit:
```suggestion
  the input tensor has already been evenly sharded on the last dimension.
```

awgu commented on docs/source/distributed.tensor.parallel.rst:
```diff
@@ -42,6 +42,11 @@ the ``parallelize_plan`` of ``parallelize_module``:
   :undoc-members:
 
+.. note:: when using the ``Shard(dim)`` as the input/output layouts for the above
+  ParallelStyles, we assume the input/output activation tensors are evenly sharded on
```
nit: 
```suggestion
  ``ParallelStyle`` s, we assume the input/output activation tensors are evenly sharded on
```

awgu commented on docs/source/distributed.tensor.parallel.rst:
```diff
@@ -42,6 +42,11 @@ the ``parallelize_plan`` of ``parallelize_module``:
 
+.. note:: when using the ``Shard(dim)`` as the input/output layouts for the above
+  ParallelStyles, we assume the input/output activation tensors are evenly sharded on
+  the tensor dimension ``dim`` on the DeviceMesh that TP operates on. For instance,
```
nit:
```suggestion
  the tensor dimension ``dim`` on the ``DeviceMesh`` that TP operates on. For instance,
```

awgu commented on docs/source/distributed.tensor.parallel.rst:
```diff
@@ -42,6 +42,11 @@ the ``parallelize_plan`` of ``parallelize_module``:
+.. note:: when using the ``Shard(dim)`` as the input/output layouts for the above
+  ParallelStyles, we assume the input/output activation tensors are evenly sharded on
+  the tensor dimension ``dim`` on the DeviceMesh that TP operates on. For instance,
+  ``RowwiseParallel`` accepts input that is sharded on the last dimension, it assumes
```
nit:
```suggestion
  since ``RowwiseParallel`` accepts input that is sharded on the last dimension, it assumes
```

awgu reviewed (approved):
Thanks for adding these clarifications!

I left some very minor stylistic nits.

wanchaol committed:
comments on "[tp] add clarification to doc and improve TP examples"


This PR adds a clarification about evenly sharded assumption in the main
tp doc and improved the tp examples by adding device mesh constructions

fixes https://github.com/pytorch/pytorch/issues/100044

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

wanchaol commented:
@pytorchbot merge

wconstab commented on docs/source/distributed.tensor.parallel.rst:
```diff
@@ -42,6 +42,14 @@ the ``parallelize_plan`` of ``parallelize_module``:
+  ``ParallelStyle``s, we assume the input/output activation tensors are evenly sharded on
+  the tensor dimension ``dim`` on the ``DeviceMesh`` that TP operates on. For instance,
+  since ``RowwiseParallel`` accepts input that is sharded on the last dimension, it assumes
+  the input tensor has already been evenly sharded on the last dimension. If needs uneven
```
nit- if needs

wanchaol committed:
Update on "[tp] add clarification to doc and improve TP examples"


This PR adds a clarification about evenly sharded assumption in the main
tp doc and improved the tp examples by adding device mesh constructions

fixes https://github.com/pytorch/pytorch/issues/100044

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

wanchaol commented:
@pytorchbot merge
